### PR TITLE
OVDB-50: Resolve mutex contention when appending AttributeArrays in parallel

### PR DESCRIPTION
**Pull request by danrbailey**

### CODE REVIEW

File: openvdb/points/AttributeArray.cc [hunk 1]
```diff
@@ -52,81 +52,71 @@ struct LockedAttributeRegistry
...
-
 // Global function for accessing the registry
 LockedAttributeRegistry*
 getAttributeRegistry()
 {
-    tbb::spin_mutex::scoped_lock lock(sInitAttributeRegistryMutex);
-
-    static LockedAttributeRegistry* registry = nullptr;
+    static LockedAttributeRegistry registry;
+    return &registry;
```
Comment (danrbailey) [inline openvdb/points/AttributeArray.cc]: @mortenbh - I took this opportunity to try and incorporate your contribution to resolve this static pointer memory leak, thanks!

Comment (mortenbh) [reply openvdb/points/AttributeArray.cc]: @danrbailey That's great! Any chance I could work with you to incorporate the rest of my changes as well? :)

File: openvdb/points/AttributeSet.cc [hunk 2]
```diff
@@ -101,7 +102,7 @@ AttributeSet::AttributeSet(const AttributeSet& attrSet, Index arrayLength)
         const Index stride = constantStride ? existingArray->stride() : existingArray->dataSize();
```
Comment (Idclip) [inline openvdb/points/AttributeSet.cc]: Is it worth creating the lock here if it's null?

File: openvdb/points/AttributeSet.cc [hunk 3]
```diff
@@ -112,13 +113,15 @@ AttributeSet::AttributeSet(const AttributeSet& attrSet, Index arrayLength)
 }
 
 
-AttributeSet::AttributeSet(const DescriptorPtr& descr, Index arrayLength)
+AttributeSet::AttributeSet(const DescriptorPtr& descr, Index arrayLength,
+    const AttributeArray::ScopedRegistryLock* lock)
     : mDescr(descr)
     , mAttrs(descr->size(), AttributeArray::Ptr())
 {
     for (const auto& namePos : mDescr->map()) {
```
Comment (Idclip) [inline openvdb/points/AttributeSet.cc]: Same here, worth creating outside the loop?

File: openvdb/points/PointAttribute.h [hunk 4]
```diff
@@ -550,11 +413,16 @@ inline void dropAttributes( PointDataTreeT& tree,
 
     // insert attributes using the new descriptor
 
-    Descriptor::Ptr newDescriptor = descriptor.duplicateDrop(indices);
+    auto newDescriptor = descriptor.duplicateDrop(indices);
```
Comment (Idclip) [inline openvdb/points/PointAttribute.h]: lots of auto :no_mouth:

File: openvdb/points/PointAttribute.h [hunk 5]
```diff
@@ -550,11 +413,16 @@ inline void dropAttributes( PointDataTreeT& tree,
...
 
-    LeafManagerT leafManager(tree);
-    tbb::parallel_for(leafManager.leafRange(),
-        DropAttributesOp<PointDataTreeT>(indices, newDescriptor));
+    tree::LeafManager<PointDataTreeT> leafManager(tree);
+    leafManager.foreach(
+        [&](typename PointDataTree::LeafNodeType& leaf, size_t /*idx*/) {
+
```
Comment (Idclip) [inline openvdb/points/PointAttribute.h]: remove nl?

File: openvdb/points/PointDelete.h [hunk 6]
```diff
@@ -127,8 +127,10 @@ struct DeleteByFilterOp
     using LeafNodeT = typename PointDataTreeT::LeafNodeType;
     using ValueType = typename LeafNodeT::ValueType;
 
-    DeleteByFilterOp(const FilterT& filter)
-        : mFilter(filter) { }
+    DeleteByFilterOp(const FilterT& filter,
+                     const AttributeArray::ScopedRegistryLock* lock)
```
Comment (Idclip) [inline openvdb/points/PointDelete.h]: perhaps default this to a nullptr to preserve api?

Comment (Idclip) [reply openvdb/points/PointDelete.h]: although I guess it's internal so perhaps not so important

**Comment by Idclip** [comment]
Really minor comments, I like the approach used here, nice work!

File: openvdb/points/AttributeSet.cc (hunk 2 above)
Comment (danrbailey) [reply openvdb/points/AttributeSet.cc]: Yes, that's a good idea.

File: openvdb/points/AttributeArray.cc (hunk 1 above)
Comment (danrbailey) [reply openvdb/points/AttributeArray.cc]: We discussed this in the last TSC meeting and I believe Jeff has replied to your PR now. :)

**Pull request merged**